tests/eve_tests/unit_tests/test_datamodels.py

```diff
@@ -0,0 +1,842 @@
+        assert issubclass(cls, ModelWithRootValidators)
+        assert isinstance(instance, ModelWithRootValidators)
+        cls.class_counter = 0
+
+    @datamodels.root_validator
+    def _another_root_validator(cls, instance):
+        assert cls.class_counter == 0
+        cls.class_counter += 1
+
+    @datamodels.root_validator
+    def _final_root_validator(cls, instance):
+        assert cls.class_counter == 1
+        cls.class_counter += 1
+
+        if instance.int_value == instance.float_value:
+            raise ValueError("'int_value' and 'float_value' must be different")
+
+
+class ChildModelWithRootValidators(ModelWithRootValidators):
+    pass
+
+
+@pytest.mark.parametrize("model_class", [ModelWithRootValidators, ChildModelWithRootValidators])
+def test_root_validators(model_class: Type[datamodels.DataModelTp]):
+    model_class(int_value=0, float_value=1.1, str_value="")
+    with pytest.raises(ValueError, match="float_value"):
+        model_class(int_value=1, float_value=1.0, str_value="")
+
+
+def test_root_validators_in_subclasses():
+    class Model(ModelWithRootValidators):
+        @datamodels.root_validator
+        def _root_validator(cls, instance):
+            assert cls.class_counter == 2
+            cls.class_counter += 10
+
+        @datamodels.root_validator
+        def _another_root_validator(cls, instance):
+            assert cls.class_counter == 12
+            cls.class_counter += 10
+
+        @datamodels.root_validator
+        def _final_root_validator(cls, instance):
+            assert cls.class_counter == 22
+            if str(instance.int_value) == instance.str_value:
+                raise ValueError("'int_value' and 'str_value' must be different")
+
+    with pytest.raises(ValueError, match="str_value"):
+        Model(int_value=1, float_value=0.0, str_value="1")
+    with pytest.raises(ValueError, match="float_value"):
+        Model(int_value=1, float_value=1.0, str_value="1")
+    with pytest.raises(ValueError, match="float_value"):
+        Model(int_value=1, float_value=1.0, str_value="1")
+
+
+# Test field options
+def test_field_init():
+    @datamodels.datamodel
+    class Model:
+        hidden_value: int = datamodels.field(init=False)
+        value: int
+        hidden_value_with_default: int = datamodels.field(default=10, init=False)
+
+    model = Model(value=1)
+    assert model.value == 1
+    assert model.hidden_value_with_default == 10
+
+    model.hidden_value = -1
+    assert model.hidden_value == -1
+
+
+def test_field_metadata():
+    @datamodels.datamodel
+    class Model:
+        value: int = datamodels.field(metadata={"my_metadata": "META"})
+
+    assert isinstance(Model.__datamodel_fields__.value.metadata, types.MappingProxyType)
+    assert Model.__datamodel_fields__.value.metadata["my_metadata"] == "META"
+
+
+# Test datamodel options
+def test_frozen():
+    import attr
+
+    @datamodels.datamodel(frozen=True)
+    class FrozenModel:
```

Comment:
What is the behavior for a frozen model with a mutable model as a field, is that allowed?

Reply:
Yes, right now the behavior is the same as dataclass/attrs, that is, the instance becomes immutable but not its fields (just like a tuple of lists or other mutable items). In the future I plan to add automatic converters to also freeze the supported collection types (sequences, sets and mappings).